DexPriceAggregatorUniswapV3/contracts/DexPriceAggregatorUniswapV3.sol

```diff
@@ -0,0 +1,409 @@
+     * Management functions *
+     ************************/
+
+    /// @notice Override the Uniswap V3 pool queried on a tokenA:tokenB route
+    /// @dev Override can be reset by using address(0) for _pool
+    /// @param _tokenA Address of an ERC20 token contract
+    /// @param _tokenB Address of another ERC20 token contract
+    /// @param _pool Address of a Uniswap V3 pool constructed with _tokenA and _tokenB
+    function setPoolForRoute(
+        address _tokenA,
+        address _tokenB,
+        address _pool
+    ) public onlyOwner {
+        PoolAddress.PoolKey memory poolKey = PoolAddress.getPoolKey(
+            _tokenA,
+            _tokenB,
+            uint24(0) // pool fee is unused
+        );
+        if (_pool != address(0)) {
+            require(
+                poolKey.token0 == IUniswapV3Pool(_pool).token0() &&
+                    poolKey.token1 == IUniswapV3Pool(_pool).token1(),
+                "Tokens or pool not correct"
+            );
+        }
+        overriddenPoolForRoute[_identifyRouteFromPoolKey(poolKey)] = _pool;
+        emit PoolForRouteSet(poolKey.token0, poolKey.token1, _pool);
+    }
+
+    /// @notice Fetch the Uniswap V3 pool to be queried for a tokenA:tokenB route
+    /// @param _tokenA Address of an ERC20 token contract
+    /// @param _tokenB Address of another ERC20 token contract
+    /// @return pool Address of a Uniswap V3 pool constructed with _tokenA and _tokenB
+    function getPoolForRoute(address _tokenA, address _tokenB)
+        public
+        view
+        returns (address pool)
+    {
+        return
+            _getPoolForRoute(
+                PoolAddress.getPoolKey(_tokenA, _tokenB, defaultPoolFee)
+            );
+    }
+
+    /**************************
+     * Utility view functions *
+     **************************/
+
+    /// @notice Fetch a Uniswap V3 pool's current "spot" and TWAP tick values
+    /// @param _pool Address of a Uniswap V3 pool
+    /// @param _twapPeriod Number of seconds in the past to consider for the TWAP rate
+    /// @return spotTick The pool's current "spot" tick
+    /// @return twapTick The twap tick for the last _twapPeriod seconds
+    function fetchCurrentTicks(address _pool, uint32 _twapPeriod)
+        public
+        view
+        returns (int24 spotTick, int24 twapTick)
+    {
+        spotTick = OracleLibrary.getBlockStartingTick(_pool);
+        twapTick = OracleLibrary.consult(_pool, _twapPeriod);
+    }
+
+    /// @notice Given a tick and a token amount, calculates the amount of token received in exchange
+    /// @param _tokenIn Address of an ERC20 token contract to be converted
+    /// @param _amountIn Amount of tokenIn to be converted
+    /// @param _tokenOut Address of an ERC20 token contract to convert into
+    /// @param _tick Tick value representing conversion ratio between _tokenIn and _tokenOut
+    /// @return amountOut Amount of _tokenOut received for _amountIn of _tokenIn
+    function getQuoteAtTick(
```

Comment:
- _tokenIn and _tokenOut should not be zero address
- _tokenIn should not be equal _tokenOut